packages/charting/src/components/LineChart/LineChart.base.tsx

```diff
@@ -3,65 +3,156 @@ import { max as d3Max } from 'd3-array';
 import { axisLeft as d3AxisLeft, axisBottom as d3AxisBottom, Axis as D3Axis } from 'd3-axis';
 import { scaleBand as d3ScaleBand, scaleLinear as d3ScaleLinear } from 'd3-scale';
 import { select as d3Select } from 'd3-selection';
+import { ILegend, Legends } from '../Legends/index';
 import { classNamesFunction } from 'office-ui-fabric-react/lib/Utilities';
-import { IProcessedStyleSet } from 'office-ui-fabric-react/lib/Styling';
-import { ILineChartProps, ILineChartStyleProps, ILineChartStyles, IDataPoint } from './LineChart.types';
+import { IProcessedStyleSet, IPalette } from 'office-ui-fabric-react/lib/Styling';
+import {
+  ILineChartProps,
+  ILineChartStyleProps,
+  ILineChartStyles,
+  IDataPoint,
+  ILineChartPoints
+} from './LineChart.types';
+import { Callout, DirectionalHint } from 'office-ui-fabric-react/lib/Callout';
 
 const getClassNames = classNamesFunction<ILineChartStyleProps, ILineChartStyles>();
 type numericAxis = D3Axis<number | { valueOf(): number }>;
 type stringAxis = D3Axis<string>;
 
-export class LineChartBase extends React.Component<ILineChartProps, {}> {
-  private _points: IDataPoint[][];
-  private _width: number;
-  private _height: number;
+export class LineChartBase extends React.Component<
+  ILineChartProps,
+  {
+    _width: number;
+    _height: number;
+    isCalloutVisible: boolean;
+    hoverValue: string | number | Date | null;
+    activeLine: string;
+    activeLegend: string;
+    lineColor: string;
+  }
+> {
+  private _points: ILineChartPoints[];
   private _lineWidth: number;
   private _strokeWidth: number;
-  private _yAxisTickCount: number;
   private _colors: string[];
   private _classNames: IProcessedStyleSet<ILineChartStyles>;
+  private _lineRootElem: HTMLElement | null;
+  private _uniqLineText: string;
 
-  public render(): JSX.Element {
-    this._adjustProps();
+  constructor(props: ILineChartProps) {
+    super(props);
+    this.state = {
+      _width: this.props.width || 600,
+      _height: this.props.height || 350,
+      isCalloutVisible: false,
+      hoverValue: '',
+      activeLine: '',
+      activeLegend: '',
+      lineColor: ''
+    };
+    this._points = this.props.data!.lineChartData || [];
+    this._lineWidth = 15;
+    this._strokeWidth = this.props.strokeWidth || 4;
+    this._uniqLineText =
+      '_line_' +
+      Math.random()
+        .toString(36)
+        .substring(7);
+
+    const { theme } = this.props;
+    const { palette } = theme!;
+    this._colors = this.props.colors || [palette.blueLight, palette.blue, palette.blueMid, palette.red, palette.black];
+    this._hoverOff = this._hoverOff.bind(this);
+  }
 
-    const isNumeric = this._points.length > 0 && this._points[0].length > 0 && typeof this._points[0][0].x === 'number';
+  public componentDidMount(): void {
```

Comment:
I'd recommend moving this code to a function passed in to the ref where you're setting `this._lineRootElem` rather than defining a componentDidMount method (similar to the _setViewBox method).

Reply:
Initially i have gone though the same method , but as per react rules we can't write setState in render because if we write it will go to infinite loop.

Won't fix.